src/Orleans.Transactions/Abstractions/TransactionalExtensionExtensions.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+
+namespace Orleans.Transactions.Abstractions
+{
+    public static class TransactionalExtensionExtensions
+    {
+        public static ITransactionalResource AsTransactionalResource(this ITransactionalExtension transactionalExtension, string resourceId)
+        {
+            return new TransactionalResourceExtensionWrapper(transactionalExtension, resourceId);
+        }
+
+        [Serializable]
+        [Immutable]
```

Comment:
is this correct? the implementation doesn't look immutable, since it's mutating the wrapped extension, and looking at the implementation for it, it's neither immutable nor [Serializable].
If this is meant to never be serialized and used inside a single silo instance, then I'd recommend that you implement a class and apply the SerializerAttribute to it, and explicitly throw on Serialize, and just pass through when (deep)copying the object, to achieve a similar effect as you get with Immutable, but with the other false behaviors.
Then remove both Serializable and Immutable attributes from this class.
You can achieve something similar without an external class by using the CopierMethodAttribute (and SerializerMethod and DeserializerMethod attributes)

Reply:
Immutability:
In this case the ITransactionalExtension is a grain reference, so all of the state is immutable.  While this is not enforced in the interface, due to how extensions are installed, this can be misused, but as it's not a commonly needed behavior I don't think their is much risk in that.

Serializability:
I'm open to other approaches, but currently I'm forcing the generation of a serializer via GenerateSerializer in the assembly info.  This was my understanding of what our recommended approach was for types that are not explicitly references in a grain interface.  Am I mistaken, is this not the preferred approach?

Reply:
You are not mistaken. It was my missunderstanding of ITransactionalExtension. I thought it was not immutable, but a grain reference is. So I thought you were not really serializing this, but just wanted to avoid deep copying.
Clarified now, ignore the comment, this approach is correct.